### PR TITLE
LPS-184830 Compare Runs Headless endpoint

### PR DESCRIPTION
**Pull request by velosofelipe64**
https://issues.liferay.com/browse/LPS-184830

### CODE REVIEW

**Comment by velosofelipe64** [comment]
ci:test:sf

File: modules/dxp/apps/osb/osb-testray/osb-testray-rest-impl/src/main/java/com/liferay/osb/testray/rest/internal/resource/v1_0/CompareRunsResourceImpl.java [hunk 1]
```diff
@@ -33,18 +36,33 @@ public class CompareRunsResourceImpl extends BaseCompareRunsResourceImpl {
 	public CompareRuns getCompareRuns(Long idTestrayRunA, Long idTestrayRunB)
 		throws Exception {
 
+		String[] dueStatuses = {
+			"PASSED", "FAILED", "BLOCKED", "TEST FIX", "DNR"
+		};
+
+		int[][] matrix = new int[5][5];
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-rest-impl/src/main/java/com/liferay/osb/testray/rest/internal/resource/v1_0/CompareRunsResourceImpl.java:43]: statusMatrix

File: modules/dxp/apps/osb/osb-testray/osb-testray-rest-impl/src/main/java/com/liferay/osb/testray/rest/internal/resource/v1_0/CompareRunsResourceImpl.java [hunk 2]
```diff
@@ -33,18 +36,33 @@ public class CompareRunsResourceImpl extends BaseCompareRunsResourceImpl {
...
+		int i;
+		int j;
+
+		for (i = 0; i < dueStatuses.length; i = i + 1) {
+			for (j = 0; j < dueStatuses.length; j = j + 1) {
+				int entry = _compareRunsService.getComparison(
+					idTestrayRunA, idTestrayRunB, dueStatuses[i],
+					dueStatuses[j], contextCompany.getCompanyId());
+
+				matrix[i][j] = entry;
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-rest-impl/src/main/java/com/liferay/osb/testray/rest/internal/resource/v1_0/CompareRunsResourceImpl.java:53]: use as inline

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 3]
```diff
@@ -14,18 +14,183 @@
...
 @Component(
 	property = "model.class.name=com.liferay.osb.testray.model.CompareRuns",
 	service = AopService.class
 )
 public class CompareRunsLocalServiceImpl
 	extends CompareRunsLocalServiceBaseImpl {
+
+	public int getComparison(
+		long runIdA, long runIdB, String statusA, String statusB,
+		long companyId) {
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:54]: sort

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 4]
```diff
@@ -14,18 +14,183 @@
...
+
+		ObjectDefinition caseObjectDefinition = _getObjectDefinitionByTableName(
+			_CASE, companyId);
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:57]: use as inline

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 5]
```diff
@@ -14,18 +14,183 @@
...
 	service = AopService.class
 )
 public class CompareRunsLocalServiceImpl
 	extends CompareRunsLocalServiceBaseImpl {
+
+	public int getComparison(
+		long runIdA, long runIdB, String statusA, String statusB,
+		long companyId) {
+
+		ObjectDefinition caseObjectDefinition = _getObjectDefinitionByTableName(
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:56]: rename to testrayCaseObjectDefinition

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 6]
```diff
@@ -14,18 +14,183 @@
...
+		Column<?, ?> caseIdColumn = table1.getColumn("c_caseId_");
+
+		DSLQuery table2 = genericQuery.where(predicateB);
+
+		return _objectEntryLocalService.dslQueryCount(
+			DSLQueryFactoryUtil.count(
+			).from(
+				table1
+			).where(
+				caseIdColumn.in(table2)
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:139]: use as inline

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 7]
```diff
@@ -14,18 +14,183 @@
...
+		Collection<Column<?, ?>> tableTemplate = new ArrayList<Column<?, ?>>() {
+		};
+
+		Column<DynamicObjectDefinitionTable, Long> caseIdDynamicColumn =
+			(Column<DynamicObjectDefinitionTable, Long>)
+				caseExtensionDynamicTable.getColumn("c_caseId_");
+
+		tableTemplate.add(caseIdDynamicColumn);
+
+		Table table1 = genericQuery.where(
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:124]: rename to table

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 8]
```diff
@@ -14,18 +14,183 @@
...
+			predicateA
+		).as(
+			"table1", tableTemplate
+		);
+
+		Column<?, ?> caseIdColumn = table1.getColumn("c_caseId_");
+
+		DSLQuery table2 = genericQuery.where(predicateB);
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:132]: use as inline

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 9]
```diff
@@ -14,18 +14,183 @@
...
+		};
+
+		Column<DynamicObjectDefinitionTable, Long> caseIdDynamicColumn =
+			(Column<DynamicObjectDefinitionTable, Long>)
+				caseExtensionDynamicTable.getColumn("c_caseId_");
+
+		tableTemplate.add(caseIdDynamicColumn);
+
+		Table table1 = genericQuery.where(
+			predicateA
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:125]: use as inline

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 10]
```diff
@@ -14,18 +14,183 @@
...
+			dueStatucColumn.eq(statusA)
+		);
+
+		Predicate predicateB = runToCaseResultColumn.eq(
+			runIdB
+		).and(
+			dueStatucColumn.eq(statusB)
+		);
+
+		JoinStep genericQuery = DSLQueryFactoryUtil.select(
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:97]: rename to joinStep

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 11]
```diff
@@ -14,18 +14,183 @@
...
+			_CASE, companyId);
+
+		DynamicObjectDefinitionTable caseExtensionDynamicTable =
+			_getDynamicObjectDefinitionTable(caseObjectDefinition, true);
+
+		ObjectDefinition caseResultObjectDefinition =
+			_getObjectDefinitionByTableName(_CASE_RESULT, companyId);
+
+		DynamicObjectDefinitionTable caseResultExtensionDynamicTable =
+			_getDynamicObjectDefinitionTable(caseResultObjectDefinition, true);
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:66]: use as inline

**Comment by velosofelipe64** [comment]
ci:test:sf

**Comment by velosofelipe64** [comment]
ci:test:sf

File: modules/dxp/apps/osb/osb-testray/osb-testray-rest-impl/src/main/java/com/liferay/osb/testray/rest/internal/resource/v1_0/CompareRunsResourceImpl.java [hunk 12]
```diff
@@ -33,18 +36,31 @@ public class CompareRunsResourceImpl extends BaseCompareRunsResourceImpl {
...
 
+		String[] dueStatuses = {
+			"PASSED", "FAILED", "BLOCKED", "TEST FIX", "DNR"
+		};
+
+		int[][] statusMatrix = new int[5][5];
+		int i;
+		int j;
+
+		for (i = 0; i < dueStatuses.length; i = i + 1) {
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-rest-impl/src/main/java/com/liferay/osb/testray/rest/internal/resource/v1_0/CompareRunsResourceImpl.java:47]: try to refator to int i = 0; ...; i++

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 13]
```diff
@@ -14,18 +14,174 @@
...
  */
 @Component(
 	property = "model.class.name=com.liferay.osb.testray.model.CompareRuns",
 	service = AopService.class
 )
 public class CompareRunsLocalServiceImpl
 	extends CompareRunsLocalServiceBaseImpl {
+
+	public int getComparison(
+		long companyId, long runIdA, long runIdB, String statusA,
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:51]: rename to testrayRunId1 and testrayRunId2, testrayDueStatus1 and testrayDueStatus2

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 14]
```diff
@@ -14,18 +14,174 @@
...
+			).eq(
+				testrayCaseToCaseResultIdColumn
+			)
+		).innerJoinON(
+			testrayCaseResultDynamicTable,
+			testrayCaseResultExtensionDynamicTable.getPrimaryKeyColumn(
+			).eq(
+				testrayCaseResultDynamicTable.getPrimaryKeyColumn()
+			)
+		);
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:100]: extract to method _getJoinStep

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 15]
```diff
@@ -14,18 +14,174 @@
...
+
+		Column<DynamicObjectDefinitionTable, String> testrayDueStatusColumn =
+			(Column<DynamicObjectDefinitionTable, String>)
+				testrayCaseResultDynamicTable.getColumn("dueStatus_");
+
+		Column<DynamicObjectDefinitionTable, Long>
+			testrayCaseToCaseResultIdColumn =
+				(Column<DynamicObjectDefinitionTable, Long>)
+					testrayCaseResultExtensionDynamicTable.getColumn(
+						"r_caseToCaseResult_c_caseId");
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:82]: put inside _getJoinStep

File: modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java [hunk 16]
```diff
@@ -14,18 +14,174 @@
...
+		tableTemplate.add(testrayCaseIdDynamicColumn);
+
+		Table table = joinStep.where(
+			testrayRunToCaseResultColumn.eq(
+				runIdA
+			).and(
+				testrayDueStatusColumn.eq(statusA)
+			)
+		).as(
+			"table1", tableTemplate
```
Comment (nNilton) [inline modules/dxp/apps/osb/osb-testray/osb-testray-service/src/main/java/com/liferay/osb/testray/service/impl/CompareRunsLocalServiceImpl.java:118]: rename to table

**Review by nNilton — CHANGES_REQUESTED** [review]

**Comment by velosofelipe64** [comment]
ci:test:sf

**Comment by velosofelipe64** [comment]
ci:test:relevant

**Comment by velosofelipe64** [comment]
ci:test:stable

**Comment by kevenleone** [comment]
ci:forward

**Pull request closed without merging**